File: 2015/03/solution.js
const solve = (visitors) => {
  let visitedHouses = {}
  let housesGifted = 0
  let positions = {}

  const visit = (position) => {
    const key = position.join(',')
    if (!visitedHouses[key]) {
      housesGifted += 1
      visitedHouses[key] = true
    }
  }

  visitors.forEach((visitor) => {
    positions[visitor] = [ 0, 0 ]
    visit(positions[visitor])
  })

  const input = require('fs').readFileSync('input.txt', 'utf8')
  for (let i = 0; i < input.length; i++) {
    const visitor = visitors[i % visitors.length]
    switch(input.charAt(i)) {
      case '^':
        positions[visitor][0] += 1
        break
      case '<':
        positions[visitor][1] -= 1
        break
      case '>':
        positions[visitor][1] += 1
        break
      case 'v':
        positions[visitor][0] -= 1
        break
    }
    visit(positions[visitor])
  }

  return housesGifted
}

console.log(`Part 1: ${solve([ 'Santa' ])} houses received at least one present.`)
console.log(`Part 2: ${solve([ 'Santa', 'Robo-Santa' ])} houses received at least one present.`)

File: 2015/03/solution.rb
class Grid

  def initialize(visitors)
    @visitors = visitors
    @visited = {}
    @gifted = 0
    @positions = {}
  end

  def solve
    @visitors.each do |visitor|
      @positions[visitor] = [ 0, 0 ]
      visit @positions[visitor]
    end
    i = 0
    File.read('input.txt').each_char do |char|
      visitor = @visitors[i % @visitors.size]
      case char
        when '^'
          @positions[visitor][0] += 1
        when '<'
          @positions[visitor][1] -= 1
        when '>'
          @positions[visitor][1] += 1
        when 'v'
          @positions[visitor][0] -= 1
      end
      visit @positions[visitor]
      i += 1
    end
    @gifted
  end

  protected
  def visit(position)
    key = position.join(',')
    if !@visited[key]
      @gifted += 1
      @visited[key] = true
    end
  end

end

puts 'Part 1', Grid.new([ 'Santa' ]).solve
puts 'Part 2', Grid.new([ 'Santa', 'Robo-Santa' ]).solve
